crates/doc/src/shape/union.rs

```diff
@@ -205,9 +205,30 @@ impl NumericShape {
             },
         ) = (lhs, rhs);
 
-        // Take the least-restrictive bounds of both.
-        let min = lhs_min.and(rhs_min).and(lhs_min.min(rhs_min));
-        let max = lhs_max.and(rhs_max).and(lhs_max.max(rhs_max));
+        // Take the least-restrictive bounds of both. We must be careful here to always prefer
+        // a float value over an integer value if the two are otherwise equal (the `Ord` impl for
+        // `Number` ignores such differences). For example, we prefer 5.0 over 5. This ensures that
+        // the minimum/maximum values are consistent with the `type` in inferred schemas.
+        let min = if let Some((lmin, rmin)) = lhs_min.zip(rhs_min) {
+            match lmin.cmp(&rmin) {
```

Comment:
nit: lift into min and max helpers, and then also use for implementing intersection ?

Reply:
The intersection logic is the inverse of this, in that it prefers integers over floats. This is to be consistent with the tightening of constraints, since the intersection of `number & integer` is `integer`. So I didn't bother factoring out a separate function.

Reply:
:+1: fair!